Add khiops-env to conda packages

bruno-at-orange opened the pull request:
Update of the khiops_env script on Linux/Mac and Windows to be compatible with conda environments (even without CONDA_PREFIX). Conda packages are updated with khiops_env (on all OS). It has an impact on packaging (cmake, NSIS) and github actions.
The important comments are in commit  a98e684f40185467c99cc9b15c3cb1754e52490c 
Add a new variable KHIOPS_MPI_HOST_FILE (in Linux only) to set up the MPI command line on multi-machines
Remove process affinity on windows

close #225

bruno-at-orange committed:
Add khiops_env in conda packages

With 3 new environment variables KHIOPS_PATH, KHIOPS_COCLUSTERING_PATH, KHIOPS_ENV_PATH, the scripts khiops,
khiops_coclustering and khiops-env are generic, we can adapt them to all plateforms. These variables are set
in the 3 scripts by cmake at the configuration step.

The khiops-core packages contains 3 more files: khiops_env, khiops and khiops-coclustering (these 2 lasts
are not mandatory but it facilitates the debugging and tests)

The MODL binary is used to get proc number in khiops_env

On UNIXes we have khiops-env, on windows khiops_env.cmd. Renaming of khiops-env to khiops_env to uniformize names between Windows and UNIXes.

Khiops_env for UNIXes
- Add function get_script_dir funtion in khiops-env.
in conda environment, the PREFIX variable is not always set. In order to set the binaries path, we use this new function.
The khiops binaries are collocated with the khiops-env script (except on fedora-like os)

NSIS
- Add replaceInfile function in NSIS, used to set the KHIOPS_HOME variable during the installation.
- NSIS code simplification

Improve khiops_env for windows
- get KHIOPS_PROC_NUMBER with command line
- add a new variable GUI which is set by cmake or nsis. This variable conditions the setting of the java related variables.
- add KHIOPS_MPI_ERROR, KHIOPS_JAVA_ERROR and KHIOPS_COCLUSTERING_PATH to fit to the UNIXes script
- We use %~dp0 to set KHIOPS_HOME, it is the directory of the script. It is no longer necessary to set KHIOPS_HOME with NSIS (or cmake) or export it.
- check mpi implementation of mpiexec

Improve khiops.cmd
- use %~dp0 to locate khiop_env.cmd (no need for KHIOPS_HOME anymore)
- use KHIOPS_MPI_ERROR and KHIOPS_JAVA_ERROR

install.cmake
- produce khiops_env.cmd
- minor fixes in UNIXes scripts
- Improve KNI package for conda: put the .dll in the bin directory and the .lib in the lib directory (as it is for KNI.zip).
- Call khiops_env -env in action test-khiops-on-iris

conda.yml
- use cmd shell for testing conda package on windows
- force process number to test in parallel

Modify khiops_env.cmd API to be the same on Windows and UNIXes
- When using the '--env' flag, variables will only be displayed without changing the environment. It is useful in debugging.

Improve the action test-khiops-install
- test khiops-env playing with KHIOPS_PROC_NUMBER

bruno-at-orange committed:
Add cmake script to detect mpi implentation (in both standard and conda environments)

bruno-at-orange committed:
Add a new binary _khiopsgetprocnumber

Add a new tiny binary to show the physical core number. It is only used by khiops_env.
This binary is packaged with khiops, so there is an impact on conda, deb, rpm and NSIS.

marcboulle reviewed (changes requested):
## Premier retour suite à installation Khiops

Installation ok (après l'obstacle de Windows Defender)

J'ai installé dans C:\My Program Files

### Test de lancement de l'outil

Problème connu: de façon standard, j'attends 20s, en mode admin, j'attend pas

Lancement de Khiops: KO
- je n'ai qu'un seul processeur disponible (selon le champ "Max number of processor cores" de l'onglet "System parameters")

### Test de lancement des scripts depuis un shell dans le répertoire bin

Remarque pour le script: %~dp0\khiops_env peut être changé en %~dp0khiops_env, car %~dp0 contient déja un caractère '\' à la fin

#### khiops

KO: j'obtiens la ligne suivante dans le shell, avant le message d'aide
'C:\My' n’est pas reconnu en tant que commande interne


#### khiops -h

KO: idem, avec la ligne suivante dans le shell
'C:\My' n’est pas reconnu en tant que commande interne


#### _getprocnumber

En lancant _getprocnumber, j'obtiens 20: ok

#### khiops_env -h

Quand je lance khiops_env -h: ok
- suggestion pour les premiere lignes:
"
Usage: khiops_env [-h, --help] [--env]
khiops_env is an internal script intended to be used by Khiops tool and Khiops wrappers only. 
It sets all the environment variables required the Khiops to run correctly (Java, MPI, etc).

options:
  -h, --help	show this help message and exit
  -env          show the environment list and exit

"    
- manque:
"KHIOPS_COCLUSTERING_PATH: path of Khiops coclustering executable, to add in path"
- correction
"number of physical cores of the computer plus one" -> "number of physical cores of the computer"

#### khiops_env --env

KO: idem, avec la ligne suivante dans le shell
'C:\My' n’est pas reconnu en tant que commande interne

Contenu affiché:
  KHIOPS_PATH C:\My Program Files\khiops\bin\MODL.exe
  KHIOPS_COCLUSTERING_PATH C:\My Program Files\khiops\bin\MODL_Coclustering.exe
  KHIOPS_MPI_COMMAND 
  KHIOPS_JAVA_PATH C:\My Program Files\khiops\jre\bin\server\;C:\My Program Files\khiops\jre\bin\
  KHIOPS_CLASSPATH C:\My Program Files\khiops\bin\khiops.jar;C:\My Program Files\khiops\bin\norm.jar
  KHIOPS_LAST_RUN_DIR C:\Users\miib6422\khiops_data\lastrun
  KHIOPS_PROC_NUMBER 1
  KHIOPS_TMP_DIR 
  KHIOPS_MEMORY_LIMIT 
  KHIOPS_API_MODE 
  KHIOPS_RAW_GUI 
  KHIOPS_JAVA_ERROR 
  KHIOPS_MPI_ERROR 

KO: deux problèmes:
- KHIOPS_MPI_COMMAND est vide
- KHIOPS_PROC_NUMBER vaut 1, et non pas 20

#### khiops_env

KO: idem, avec la ligne suivante dans le shell
'C:\My' n’est pas reconnu en tant que commande interne

Variables d'environnement positionnéees ("set | grep KHIOPS"): ok, conforme au résultats de khiops-env --env
  KHIOPS_CLASSPATH=C:\My Program Files\khiops\bin\khiops.jar;C:\My Program Files\khiops\bin\norm.jar
  KHIOPS_COCLUSTERING_PATH=C:\My Program Files\khiops\bin\MODL_Coclustering.exe
  KHIOPS_HOME=C:\My Program Files\khiops
  KHIOPS_JAVA_PATH=C:\My Program Files\khiops\jre\bin\server\;C:\My Program Files\khiops\jre\bin\
  KHIOPS_LAST_RUN_DIR=C:\Users\miib6422\khiops_data\lastrun
  KHIOPS_PATH=C:\My Program Files\khiops\bin\MODL.exe
  KHIOPS_PROC_NUMBER=1

popescu-v reviewed (commented):
Some comments (clarifications)

marcboulle reviewed (changes requested):
Je pense qu'il vaut mieux renommer _getprocnumber en _khiopsgetprocnumber pour être vraiment sûr d'être dans "name space" spécifique khiops et éviter tout conflit potentiel (même si peu probable)

popescu-v reviewed (changes requested):
See comments.

bruno-at-orange committed:
Improve khiops_env (cluster ready and more)

Add KHIOPS_MPI_HOST_FILE to khiops_env
- Allows to specify a hostfile to mpiexec
- if this variable is defined, khiops_env will export environment variables required by khiops to run correctly to MPI processes.
These variables relate to GCS/S3 credentials and to khiops itself (KHIOPS_* and Khiops*)
khiops_env is built on cmake configure: the variable export is done only for openmpi and the hostfile keyword is different for openmpi and mpich

Remove process affinity for MS-MPI
- As discussed in #350, the affinity slows down khiops.
- remove the prioryty (below normal) too

Improve UNIXes khiops_env
- java configuration and modules loading are not in khiops_env on conda environment
- To improve the install.cmake readability, 3 files are added to fill variables:
	- export_env_variables.sh
	- java_settings.sh
	- use_environment_module.sh

fix GUI error message in khiops and khiops coclustering
- The message saying to install "khiops" on conda environment is wrong.
- when a non-existing option is passed, e.g. khiops --foo, the GUI message is displayed

fix message in help about KHIOPS_API_MODE: false <=> true

KHIOPS_API_MODE message

marcboulle reviewed (dismissed):
Tests d'installation avec l'installeur Windows: OK
Prise en compte des corrections: OK

Concernant le nouvel exécutable _khiopsgetprocnumber.exe, sous Windows
- attention: il ne faudra pas oublier de le signer dans la procédure de fabrication de l'installeur Windows, dans sa version diffusable
- suggestion:  permettre d'avoir les propriétés par clic-bouton droit, comme pour les autre exe
  - version basique, avec numéro de version 1.0 indépendante de celle de Khiops
  - s'inspirer de ce qui fait dans Learning/khisto
     - khisto.rc, resource.h
     - Version.h ici non nécessaire
     - attention à adapter CMakeLists.txt

popescu-v reviewed (changes requested):
See the new comments.

bruno-at-orange committed:
Put tests results in artifacts on action test-khiops-on-iris

bug fix: when running tests without using kht_test, we have to set the variable KHIOPS_API_MODE to true

popescu-v reviewed (approved):
LGTM
